src/borg/_hashindex.c

```diff
@@ -569,11 +570,11 @@ hashindex_read(PyObject *file_py, int permit_compact, int legacy)
     }
     index->buckets = index->buckets_buffer.buf;
 
-    if(!permit_compact) {
-        index->min_empty = get_min_empty(index->num_buckets);
-        if (index->num_empty == -1)  // we read a legacy index without num_empty value
-            index->num_empty = count_empty(index);
+    index->min_empty = get_min_empty(index->num_buckets);
+    if (index->num_empty == -1)  // we read a legacy index without num_empty value
+        index->num_empty = count_empty(index);
 
+    if(!permit_compact) {
```

Comment:
I moved the setting of `min_empty` and `num_empty` to outside of the if, so they happen even when `permit_compact` is true.  I feel like otherwise they might contain garbage that could in theory cause problems.  This is unrelated to the rest of the PR, so I'll drop it if there is any doubt about it.

Reply:
Will have a deeper look at this later.

Reply:
(This is actually slight related to this PR, since I did add `permit_compact=True` to the test suite.)

Reply:
LGTM.

I just noticed that `count_empty` could be tweaked for compact hashtables (where we can detect there can be no empty buckets because `num_entries == num_buckets`).

But this doesn't need to get optimized in master, because it now persists the `num_empty` value on disk anyways.